### DISCUSSION TITLE
Feature Request: Maintainer prompting guides and claude project tool.

### DISCUSSION THREAD

**Discussion by atomicwrite** [ServiceStack/Discuss#130] [Ideas]
## **Proposal: A Comprehensive Guide for AI-Assisted ServiceStack Development**

We propose developing a comprehensive guide for leveraging AI models (such as Claude) in ServiceStack development. This guide will include example prompts, best practices, and community insights to assist developers of all experience levels in utilizing AI for their ServiceStack projects.

### **ServiceStack's Declarative Nature**

ServiceStack's declarative approach aligns well with AI-assisted development. Its clear, expressive syntax facilitates more accurate interpretation by AI models. By leveraging ServiceStack's declarative patterns in our prompts, we can guide AI to generate more idiomatic and effective code. This guide will exploit the synergy between ServiceStack's design philosophy and AI capabilities to enhance the development experience.

### **Knowing ServiceStack's Tricks:**

To get the most out of AI, it's important to understand how ServiceStack works. The people who made ServiceStack know the best ways to use it. By using their tips and tricks in our prompts, we can help AI write code that's not just correct but also looks good and follows ServiceStack's rules.

## **Leveraging VFS for Seamless AI Integration:**

* **Virtual File System (VFS) Integration:** Utilize VFS to provide a unified, abstract view of local and remote file systems. This will enable the tool to seamlessly interact with project files and structures, regardless of their physical location.
* **Claude API Integration:** Allow users to input their Claude API key, enabling the tool to access Claude's capabilities for generating code, suggestions, and refactoring recommendations.

#### **Enhanced AI-Assisted Development Features:**

* **AI-Driven Code Generation:** Enable developers to input a file or project structure and receive AI-generated suggestions for building it out. The tool can leverage Claude's understanding of ServiceStack to generate relevant code snippets, templates, or complete components.
* **Context-Aware Code Completion and Suggestions:** Provide intelligent code completion and suggestions based on the context of the current file and project structure. The tool can analyze existing code, recognize patterns, and offer relevant recommendations to improve code quality and efficiency.
* **Automated Code Generation:** Facilitate the generation of complementary files and code based on existing components. For example, the tool could automatically generate a service implementation from a service interface or create necessary configuration files based on project settings.
* **Intelligent Refactoring:** Offer intelligent refactoring suggestions to align code with ServiceStack best practices and improve code maintainability. The tool can analyze code for potential improvements, identify areas for refactoring, and suggest appropriate changes.

#### **GitHub Integration (Optional):**

* **Branch Creation and Management:** Allow users to create new GitHub branches and manage existing branches directly from within the tool. This would streamline the development workflow and make it easier to experiment with different code changes without affecting the main branch.
* **Pull Request Creation and Review:** Facilitate the creation and review of pull requests, making it easier for developers to collaborate on code changes and ensure code quality.

I have the scaffold of one and a github oauth flow that doesn't require mvc for this project. It wasn't hard and now with the prompts if they come out all of that should be quick. I found it's best to create a prompt on a real project or while making one. Maybe develop that tool for claude service / stack.

**Comment by atomicwrite** [comment]
tl;dr 

maintainer system prompts and an open version (there are tons already) of a cursor tool that you can give a scaffold to and it write it out. Like, write me a back end for a medical company that needs to store hippa compliant records encrypted on aws buckets that only updates on allowed ips but lets the data be collected offline.

**Comment by atomicwrite** [comment]
or more realistically, browse your github and say, ok claude, these files, this system prompt and this problem.

"projects" is sorta that but you still cpoy and paste and sometimes it cuts out stuff to save tokens.

**Comment by mythz** [comment]
Thanks for submitting this comprehensive feature request!

We're currently working on finishing our [AI Server Project](https://servicestack.net/posts/scalable-sqlite#ai-server) which is the gateway we're using to build all our other AI features on top of, after we finish that we'll have some spare capacity to look into this.

I'd be interested in looking at creating some guides to help support AI Assisted development with ServiceStack. It would be much smaller in scope than a comprehensive tool to manage the entire AI Assisted development workflow, instead we'd want to integrate with an existing popular (and ideally free) AI Assisted tool to interface with Claude/LLMs to make the actual changes. 

So we'd primarily be looking at instructional guides we could load into these tools to help guide them generate correct ServiceStack APIs. I agree we should focus on generating declarative APIs, e.g. AutoQuery CRUD APIs and ideally a [Custom UI](https://servicestack.net/posts/admin-uis) which IMO stands the best chance to generate a working UI with backend APIs. 
The great thing is that we control the entire stack which will allow us to add extra [Declarative Attributes](https://docs.servicestack.net/locode/declarative) or [Vue Components](https://docs.servicestack.net/vue/) if it can help.

The first thing would be focus on a specific selection of tools that we would recommend and test against, so it would be good to know which tools you're using? So we can use the same tools. 

IMO these are the popular choices:

IDEs:
 - VS Code
 - Rider

AI Assisted Tools:
 - Aider
 - Claude Dev
 - Continue.dev

Please let me know which other tools you use and would recommend.

If we can manage to get a good dev workflow, one feature we could look at creating is an "online project creator" where you could specify a prompt like the type of App you would like to build, which we could try to use to create a new Project with the Data Models, AutoQuery CRUD APIs and UI. But wont really know how effective the result will be until we build it and experiment with different style of prompts.

**Comment by atomicwrite** [comment]
That's impressive. I like to talk to claude before I build. I did some examples with my prompt and it doesn't 100% annotate the right things. It sometimes ties to group icreate,iupdate on the same class. What would help is just a very detailed guide as if you were talking to a 4 year old about how each piece works. I'll update mine as I find quirks like these but I know 1/10 of this frame work and that's after using it for years.

The image and video pipeline looks neat. I'm trying continue dev since it is the only one that works with rider.

**Comment by atomicwrite** [comment]
Wrote a quick POC for what the tooling I'd like to see in your AI stuff. https://github.com/atomicwrite/Clood Basically, these files, this prompt, review. It auto branches and commits. Feel free to use any of it.

**Reply by mythz** [reply]
Cool, not sure on the approach we'd look at taking just yet but didn't plan on git integration, ideally we'd just be focused on generating new APIs/UI classes/files and add them to existing projects (similar to how the [mix tool](https://docs.servicestack.net/mix-tool) adds gist files to projects), IMO other existing tools would be better served at loading context / modifying existing code bases, etc. Our main focus would be primarily on the new code generation.

**Comment by mythz** [comment]
FYI we've just [released ServiceStack v8.6](https://docs.servicestack.net/releases/v8_06) with our initial support for using AI to generate CRUD APIs and UI from a text prompt.

You can generate a new Blazor CRUD App from a text prompt at:
### https://servicestack.net/text-to-blazor

The new `okai` tool can be used to add AutoQuery CRUD APIs to an existing project with:

```sh
npx okai <prompt>
```

You can find more info about the tools in the [ServiceStack v8.6 Release Notes](https://docs.servicestack.net/releases/v8_06).
